### PR TITLE
Strings methods: `find()`

### PR DESCRIPTION
**Pull request by Abdelrahman-Kh-Fouad**

### CODE REVIEW

File: src/lpython/semantics/python_ast_to_asr.cpp [hunk 1]
```diff
@@ -4356,6 +4356,65 @@ class BodyVisitor : public CommonVisitor<BodyVisitor> {
...
+                    if (args.size() != 1) {
+                        throw SemanticError("str.find() takes one arguments",
+                            x.base.base.loc);
+                    }
+                    ASR::expr_t *arg = args[0].m_value;
+                    ASR::ttype_t *type = ASRUtils::expr_type(arg);
+                    if (ASRUtils::is_character(*type)) {
+                        std::string str = (ASR::down_cast<ASR::StringConstant_t>(at->m_value))->m_s;
+                        if (ASRUtils::expr_value(arg) != nullptr) {
+                            ASR::StringConstant_t* ss = ASR::down_cast<ASR::StringConstant_t>(ASRUtils::expr_value(arg));
```
Comment (Abdelrahman-Kh-Fouad) [inline src/lpython/semantics/python_ast_to_asr.cpp:4369]: @certik @czgdp1807 line i mentioned on zulip

Comment (Thirumalai-Shaktivel) [reply src/lpython/semantics/python_ast_to_asr.cpp:4369]: Above at line number `4367`: `at->value` is of type `LFortran::LPython::AST:expr_t *`. I think we have downcast it using `AST::down_cast<AST::ConstantStr_t>(at->m_value)`, just like we did in line `4348`

Comment (Abdelrahman-Kh-Fouad) [reply src/lpython/semantics/python_ast_to_asr.cpp:4369]: @Thirumalai-Shaktivel 
i think casting works with AST and ASR, i tried them and line 4369 still raise the same error
```bash
In file included from /home/abdelrahmankh/Gsoc/lpython/src/libasr/alloc.h:10,
                 from /home/abdelrahmankh/Gsoc/lpython/src/libasr/asr.h:6,
                 from /home/abdelrahmankh/Gsoc/lpython/src/lpython/semantics/python_ast_to_asr.cpp:13:
/home/abdelrahmankh/Gsoc/lpython/src/lpython/python_ast.h: In instantiation of ‘T* LFortran::LPython::AST::down_cast(const U*) [with T = LFortran::ASR::StringConstant_t; U = LFortran::ASR::expr_t]’:
/home/abdelrahmankh/Gsoc/lpython/src/lpython/semantics/python_ast_to_asr.cpp:4369:94:   required from here
/home/abdelrahmankh/Gsoc/lpython/src/lpython/python_ast.h:40:28: error: call of overloaded ‘is_a<LFortran::ASR::StringConstant_t>(const LFortran::ASR::expr_t&)’ is ambiguous
   40 |     LFORTRAN_ASSERT(is_a<T>(*f));

```

Comment (Abdelrahman-Kh-Fouad) [reply src/lpython/semantics/python_ast_to_asr.cpp:4369]: yes, it works now, thanks @Thirumalai-Shaktivel .

File: src/runtime/lpython_builtin.py [hunk 2]
```diff
@@ -638,6 +638,51 @@ def _lpython_str_lower(x: str) -> str:
...
+    sub_len :i32
+    s_len = len(s)
+    sub_len = len(sub)
+    flag: bool
+    flag = False
+    res: i32
+    res = -1
+    lps: list[i32]
+    i :i32
+    len: i32
```
Comment (czgdp1807) [inline src/runtime/lpython_builtin.py:652]:
```suggestion
    s_len :i32; sub_len :i32; flag: bool;
    res: i32; i: i32; len: i32;
    lps: list[i32]

    s_len = len(s)
    sub_len = len(sub)
    flag = False
    res = -1
```

File: src/runtime/lpython_builtin.py [hunk 3]
```diff
@@ -638,6 +638,51 @@ def _lpython_str_lower(x: str) -> str:
...
+    for i in range(sub_len):
+        lps.append(0)
```
Comment (czgdp1807) [inline src/runtime/lpython_builtin.py:654]:
```suggestion
        lps.append(0)

```

File: src/runtime/lpython_builtin.py [hunk 4]
```diff
@@ -638,6 +638,51 @@ def _lpython_str_lower(x: str) -> str:
...
+        if sub[i] == sub[len]:
+            len += 1
+            lps[i] = len
+            i += 1
+        else:
+            if len != 0:
+                len = lps[len - 1]
+            else:
+                lps[i] = 0
+                i += 1
```
Comment (czgdp1807) [inline src/runtime/lpython_builtin.py:667]:
```suggestion
                i += 1

```

File: src/runtime/lpython_builtin.py [hunk 5]
```diff
@@ -638,6 +638,51 @@ def _lpython_str_lower(x: str) -> str:
...
+            j += 1
+        if j == sub_len:
+            res = i- j
+            flag = True
+            j = lps[j - 1]
+        elif i < s_len and sub[j] != s[i]:
+            if j != 0:
+                j = lps[j - 1]
+            else:
+                i = i + 1
```
Comment (czgdp1807) [inline src/runtime/lpython_builtin.py:683]:
```suggestion
                i = i + 1

```

File: integration_tests/test_str_attributes.py [hunk 6]
```diff
@@ -25,6 +25,17 @@ def strip():
     assert "     AASAsaSas    " .lstrip() == "AASAsaSas    "
     assert "     AASAsaSas    " .strip() == "AASAsaSas"
 
+def find():
+    s: str
+    sub: str
+    s = "AaaaAABBbbbbBB!@12223BN"
+    sub = "@"
```
Comment (czgdp1807) [inline integration_tests/test_str_attributes.py:32]: Test for sub-strings of length more than 1? And probably empty sub-string as well?

**Review by czgdp1807 — CHANGES_REQUESTED** [review]

File: src/runtime/lpython_builtin.py (hunk 2 above)
Comment (czgdp1807) [reply src/runtime/lpython_builtin.py:652]: I would recommend to name your variable as `len_` i.e., something different from intrinsic `len`.

File: integration_tests/test_str_attributes.py (hunk 6 above)
Comment (Abdelrahman-Kh-Fouad) [reply integration_tests/test_str_attributes.py:32]: I added them

File: src/runtime/lpython_builtin.py (hunk 2 above)
Comment (Abdelrahman-Kh-Fouad) [reply src/runtime/lpython_builtin.py]: now everything is ready now.
but, i don't know what happend with windows built!

**Review by czgdp1807 — CHANGES_REQUESTED** [review]
Resolve the conflicts and clean up the git history. See https://github.com/lcompilers/lpython/blob/main/doc/src/rebasing.md#merging for steps.

**Comment by Abdelrahman-Kh-Fouad** [comment]
@czgdp1807 Done!

**Comment by Abdelrahman-Kh-Fouad** [comment]
@czgdp1807  ready now!

File: integration_tests/test_str_attributes.py [hunk 7]
```diff
@@ -25,6 +25,26 @@ def strip():
...
+    assert s.find(sub) == 15
+    assert s.find('B') == 6
+    assert "empty strings" .find("string") == 6
+    s2: str
+    s2 = "Ferrars all spirits his imagine effects amongst neither. It bachelor cheerful of mistaken. Tore has sons put upon wife use bred seen. Its dissimilar invitation ten has discretion unreserved. Had you him humoured jointure ask expenses learning. Blush on in jokes sense do do. Brother hundred he assured reached on up no. On am nearer missed lovers. To it mother extent temper figure better."
```
Comment (czgdp1807) [inline integration_tests/test_str_attributes.py:37]: What's the source of this string?

Comment (czgdp1807) [reply integration_tests/test_str_attributes.py:37]: Coming from https://www.rs-niers-mg.de/breaking-bad-illustrated/?

Well copying a string from a website makes us prone to copyright claims. Can you just write something of your own? Like just take this review comment and put it as a string?

**Review by czgdp1807 — CHANGES_REQUESTED** [review]

File: integration_tests/test_str_attributes.py (hunk 7 above)
Comment (Abdelrahman-Kh-Fouad) [reply integration_tests/test_str_attributes.py:37]: it's from a text generator, but i will change it.

Comment (Abdelrahman-Kh-Fouad) [reply integration_tests/test_str_attributes.py]: @czgdp1807 
ready now.

**Review by czgdp1807 — APPROVED** [review]

**Pull request merged**